misc/multistream-select/src/listener_select.rs

```diff
@@ -42,6 +42,37 @@ use std::{
 /// returned `Future` resolves with the name of the negotiated protocol and
 /// a [`Negotiated`] I/O stream.
 pub fn listener_select_proto<R, I>(inner: R, protocols: I) -> ListenerSelectFuture<R, I::Item>
+where
+    R: AsyncRead + AsyncWrite,
+    I: IntoIterator,
+    I::Item: AsRef<[u8]>,
+{
+    listener_select_proto_with_state(
+        State::RecvHeader {
+            io: MessageIO::new(inner),
+        },
+        protocols,
+    )
+}
+
+/// Used when selected as a [`crate::Role::Responder`] during [`crate::dialer_select_proto`]
+/// negotiation with [`crate::Version::V1SimultaneousOpen`]
+pub(crate) fn listener_select_proto_no_header<R, I>(
+    io: MessageIO<R>,
+    protocols: I,
+) -> ListenerSelectFuture<R, I::Item>
+where
+    R: AsyncRead + AsyncWrite,
+    I: IntoIterator,
+    I::Item: AsRef<[u8]>,
+{
+    listener_select_proto_with_state(State::RecvMessage { io }, protocols)
+}
+
+fn listener_select_proto_with_state<R, I>(
+    state: State<R, I::Item>,
+    protocols: I,
+) -> ListenerSelectFuture<R, I::Item>
```

Comment:
I guess you are just following the style here but I always wondered why these aren't associated constructor functions on `ListenerSelectFuture`.

```rust
impl<R, N> ListenerSelectFuture<R, N> {
	pub fn new(inner: R, protocols: I) -> Self { }
	pub fn in_state(state: State<R, I::Item>, protocols: I) -> Self { } 
	pub fn assume_header_sent(io: MessageIO<R>, protocols: I) -> Self { }
}
```